keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -557,278 +532,231 @@ reconstructed by the driver. This has no effect on already allocated claims
 ...
 spec:
-  minimumRuntimeVersion: v12.0.0
-  minimumMemory: 32Gi
-  # "sharing" is a configuration parameter that does not
-  # get translated into the selector below.
-  sharing:
-    strategy: TimeSliced
-```
-
-Note that all fields in this CRD can be fully validated since it is owned by
-the DRA driver itself. This includes value ranges that are specific to the
-underlying hardware. There's no risk of using invalid attribute names because
-only the fields shown here are valid.
-
-With this CRD in place, a DRA driver controller is able to convert instances of
-it into a generic, in-tree `ResourceClaimParameters` object that the scheduler
-is able to understand.
-
-For the example above, the converted object would look as follows:
-
-```yaml
-kind: ResourceClaimParameters
-apiVersion: resource.k8s.io/v1alpha2
-
-metadata:
-  # This cannot be the same as my-parameters because parameter objects with a different
-  # type might also use it. Instead, the original object gets linked to below.
-  name: someArbitraryName
-  namespace: user-namespace
-
-generatedFrom:
-  name: my-parameters
-  kind: CardParameters
-  apiGroup: dra.example.com
-  uid: foobar-uid
-
-vendorParameters:
-  # A vendor can put any kind of object here to pass the configuration
-  # parameters down to the kubelet plugin. In this case, the vendor
-  # driver simply copied the entire CR. It could also be some
-  # separate, smaller configuration type.
-  #
-  # Beware that ResourceClaimParameters have separate RBAC rules than
-  # the vendor CRD, so information included here may get visible
-  # to more users than the original CRD. Both objects are in the same
-  # namespace.
-  kind: CardParameters
-  apiVersion: dra.example.com/v1alpha1
-  metadata:
-    name: my-parameters
-    namespace: user-namespace
-    uid: foobar-uid
-  ...
-  spec:
-    minimumRuntimeVersion: v12.0.0
-    minimumMemory: 32Gi
-    sharing:
-      strategy: TimeSliced
-
-driverRequests:
-- driverName: cards.dra.example.com
-  requests:
-  # Each entry here is a request for one resource.
-  - namedresources:
-      # Selectors are CEL expressions with access to the attributes of the named resource
-      # that is being checked for a match.
-      selector: |-
-        attributes.version["runtimeVersion"].isGreaterThan(semver("12.0.0")) &&
-        attributes.quantity["memory"].isGreaterThan(quantity("32Gi"))
+  # The node name indicates the node.
+  # Each driver on a node provides pools of devices for allocation,
+  # with unique device names inside each pool.
+  # Usually, but not necessarily, that pool name is the same as the
+  # node name.
+  nodeName: worker-1
+  poolName: worker-1
+  driverName: gpu.dra.example.com
+  sharedCapacity:
+  - name: gpu-0-memory-block-0
+    capacity: 1
+  - name: gpu-0-memory-block-1
+    capacity: 1
+  - name: gpu-0-memory-block-2
+    capacity: 1
+  - name: gpu-0-memory-block-3
+    capacity: 1
+  devices:
+  - name: gpu-0
+    attributes:
+    - name: memory
```

Comment:
I had proposed something similar [here](https://github.com/kubernetes-sigs/wg-device-management/commit/cd5ba42134bf730d34a1c5106968beef97353af4), which unfortunately never got much attention, and was reverted by Patrick as "out of scope" for 1.31.

In that proposal I actually suggested that we remove all `Quantity` fields from `DeviceAttributes` and instead list them in a device as either (1) part of consuming some `SharedCapacity` or (2) as a `LocalCapacity`.

Users could then request these capacities in a Claim independent of the CEL expression (which is just for attributes) in a dedicated `resources` field (just like we have in the pod spec for "normal" resource requests).

e.g.
```
    - name: gpu-request
      deviceClassName: acme-gpu
      selectors:
      - expression: device.attributes["cudaComputeCapability"].isGreaterThan(semver("8.0")) 
      resources:
          memory: 40Gi
```

Expanding the CEL to have `capacity` instead of just `attributes` would work just as well though.

Reply:
Yes, I think that's better too. I prefer to keep allocatable things separate from attributes; I had this in various iterations in the past. But we pulled out allocatable because of scope.

"Allocatable Capacity" is theoretically a subset of "quantity typed". If we put all "quantity typed" attributes in "Capacity", does that actually make sense? Are there things that are quantity values, that we don't want to treat as consumable? If we don't put all quantity typed things in `capacity`, how will a user know when to use `device.attributes` and `device.capacity` in the CEL?

In the earlier prototypes, I separated it so CEL operates on attributes, and you have `requests` to allocate from `capacity`. This was sort of put aside though since we are not doing fractional devices. With only whole device allocation, it makes sense to do things like "memory >= 40Gi" to select the whole card, so we had to put memory into attributes.

I guess we could still create the *API* for partial allocation now, even if actual allocation is whole device. So if the user makes a claim with requests for "10Gi", we will still grant them the whole GPU. 

In that case, it would look a lot closer to what I expect the final shape to be (ie, claims would have a per-device resource requests field, in addition to a selector expression, amount field, and constraints on the set).

Reply:
> In that proposal I actually suggested that we remove all Quantity fields from DeviceAttributes

If we move allocatable things out of attributes, then this seems reasonable to me, though we would need to bring back `IntValue`, probably.  I doubt that there are non-capacity attributes that benefit from quantityt notation, but if that is wrong, we could bring it back.

> Users could then request these capacities in a Claim

I think that is ultimately coming, without much doubt.  I also thing it makes sense to drop it from this release.  That doesn't mean that we shouldn't start segregating "attribs" from "capacity".  In this release people can treat `capacity[]` as a 2nd set of attributes and constrain against it.  Later they make requests against it.

> I guess we could still create the API for partial allocation now, even if actual allocation is whole device. So if the user makes a claim with requests for "10Gi", we will still grant them the whole GPU.

We don't need to put the `requests` API in yet - just let people reference `dev.capacity["memory"]` in CEL.  "I want a GPU with at least 20Gi of memory, but I get the whole device".  When we add `requests` the "at least" will be implied, but you will get _only_ 20Gi.